python/dgl/sparse/sparse_matrix.py

```diff
@@ -194,6 +172,192 @@ def transpose(self):
         """
         return SparseMatrix(self.c_sparse_matrix.transpose())
 
+    def to(self, device=None, dtype=None):
+        """Perform matrix dtype and/or device conversion. If the target device
+        and dtype are already in use, the original matrix will be returned.
+
+        Parameters
+        ----------
+        device : torch.device, optional
+            The target device of the matrix if provided, otherwise the current
+            device will be used
+        dtype : torch.dtype, optional
+            The target data type of the matrix values, otherwise the current
+            data type will be used
+
+        Returns
+        -------
+        SparseMatrix
+            The result matrix
+
+        Example
+        --------
+
+        >>> row = torch.tensor([1, 1, 2])
+        >>> col = torch.tensor([1, 2, 0])
+        >>> A = from_coo(row, col, shape=(3, 4))
+        >>> A.to(device='cuda:0', dtype=torch.int32)
+        SparseMatrix(indices=tensor([[1, 1, 2],
+                                     [1, 2, 0]], device='cuda:0'),
+                     values=tensor([1, 1, 1], device='cuda:0',
+                                   dtype=torch.int32),
+                     size=(3, 4), nnz=3)
+        """
+        if device is None:
+            device = self.device
+        if dtype is None:
+            dtype = self.dtype
+
+        if device == self.device and dtype == self.dtype:
+            return self
+        elif device == self.device:
+            return val_like(self, self.val.to(dtype=dtype))
+        else:
+            row, col = self.coo()
+            row = row.to(device=device)
+            col = col.to(device=device)
+            val = self.val.to(device=device, dtype=dtype)
+            return from_coo(row, col, val, self.shape)
```

Comment:
Still don't think this is the right way. Perhaps it needs a C++ operator, which will copy/cast whatever formats exist in the sparse matrix to the target.

Reply:
I am not sure what's the difference here. And the dtype change is (likely) not a heavy ops in the training cycle, so probably not worth to optimize more?

Reply:
Discussed with @frozenbugs, we will table the optimization to post v1.0. cc @mufeili

Reply:
Filed https://github.com/dmlc/dgl/issues/5119 to track.